crates/meilisearch-types/src/keys.rs

```diff
@@ -195,206 +196,294 @@ fn parse_expiration_date(
+    fn deserialize_from_value<V: deserr::IntoValue>(
+        value: deserr::Value<V>,
+        location: deserr::ValuePointerRef<'_>,
+    ) -> Result<Self, E> {
+        match value {
+            deserr::Value::String(s) => match Self::get_action(&s) {
+                Some(action) => Ok(action),
+                None => Err(deserr::take_cf_content(E::error::<std::convert::Infallible>(
+                    None,
+                    deserr::ErrorKind::UnknownValue {
+                        value: &s,
+                        accepted: &Self::SERDE_MAP_ARR.map(|(ser_action, _)| ser_action),
+                    },
+                    location,
+                ))),
+            },
+            _ => Err(take_cf_content(E::error(
+                None,
+                deserr::ErrorKind::IncorrectValueKind {
+                    actual: value,
+                    accepted: &[deserr::ValueKind::String],
+                },
+                location,
+            ))),
+        }
+    }
+}
+
+impl Serialize for Action {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(Self::get_action_serde_name(self))
+    }
+}
+
+impl<'de> Deserialize<'de> for Action {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct Visitor;
+        impl serde::de::Visitor<'_> for Visitor {
+            type Value = Action;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                write!(formatter, "the name of a valid action (string)")
+            }
+
+            fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                match Self::Value::get_action(s) {
+                    Some(action) => Ok(action),
+                    None => Err(E::invalid_value(serde::de::Unexpected::Str(s), &"a valid action")),
+                }
+            }
+        }
+
+        deserializer.deserialize_str(Visitor)
+    }
+}
+
+// TODO: Once "all" type flags are removed, simplify
+//       Essentially `get_potential_index` will give the exact index, +1 the exact next, -1 the exact previous
+impl Sequence for Action {
+    const CARDINALITY: usize = Self::FLAGS.len();
+
+    fn next(&self) -> Option<Self> {
+        let mut potential_next_index = self.get_potential_index() + 1;
+
+        loop {
+            if let Some(next_flag) = Self::FLAGS.get(potential_next_index).map(|v| v.value()) {
+                if next_flag > self {
+                    return Some(*next_flag);
+                }
+
+                potential_next_index += 1;
+            } else {
+                return None;
+            }
+        }
+    }
+
+    fn previous(&self) -> Option<Self> {
+        // -2 because of "all" type flags that represent a single flag, otherwise -1 would suffice
```

Comment:
By removing the `ALL` flag and using the `Flags::all` method we can get rid of this conplexity.

Reply:
related to:
https://github.com/meilisearch/meilisearch/issues/5489